mxnet/mxnet-engine/src/main/java/ai/djl/mxnet/jna/ObjectPool.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
+ * with the License. A copy of the License is located at
+ *
+ * http://aws.amazon.com/apache2.0/
+ *
+ * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+ * OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
+ * and limitations under the License.
+ */
+package ai.djl.mxnet.jna;
+
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+import javax.naming.spi.ObjectFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A generic object pool implementation.
+ *
+ * @param <T> the type of object to put in the pool
+ */
+@SuppressWarnings("MissingJavadocMethod")
+public class ObjectPool<T> {
+
+    private static final Logger logger = LoggerFactory.getLogger(ObjectFactory.class);
+
+    private BlockingQueue<T> queue;
+    private Supplier<T> supplier;
+    private Consumer<T> consumer;
+
+    public ObjectPool(int size, Supplier<T> supplier, Consumer<T> consumer) {
+        queue = new ArrayBlockingQueue<>(size * getPoolSizeFactor());
+        this.supplier = supplier;
+        this.consumer = consumer;
+    }
+
+    public T acquire() {
+        T item = queue.poll();
+        if (item == null) {
+            if (supplier != null) {
+                return supplier.get();
```

Comment:
Same issues here. If system requires more PointerByReference, then supplier will be simply provide that without blocking. How will we effectively control the number?

Reply:
We don't want to control how many object is created, This will make sure no behavior change for any existing code. We only how many can be cached. If all the used objects are added to the pool, when it hit the max, no more new object will be created. I'm thinking if we should change the queue to unbounded, so we can cache unlimited objects.